pkg/distributor/ha_tracker.go

```diff
@@ -204,9 +246,112 @@ func (c *haTracker) loop(ctx context.Context) error {
+	// If selected replica is marked for deletion for this time, it is deleted completely.
+	deletionTimeout = 30 * time.Minute
+)
+
+func (c *haTracker) cleanupOldReplicasLoop(ctx context.Context) {
+	tickerTimeout := cleanupCyclePeriod + time.Duration(rand.Int63n(cleanupCycleJitter.Nanoseconds())-cleanupCycleJitter.Nanoseconds()/2)
+
+	tick := time.NewTicker(tickerTimeout)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case t := <-tick.C:
+			c.cleanupRuns.Inc()
+			c.cleanupOldReplicas(ctx, t.Add(-deletionTimeout))
+		}
+	}
+}
+
+// Replicas marked for deletion before deadline will be deleted.
+// Replicas with last-received timestamp before deadline will be marked for deletion.
+func (c *haTracker) cleanupOldReplicas(ctx context.Context, deadline time.Time) {
+	keys, err := c.client.List(ctx, "")
+	if err != nil {
+		level.Error(c.logger).Log("msg", "cleanup: failed to list replica keys", "err", err)
+		return
+	}
+
+	for _, key := range keys {
+		if ctx.Err() != nil {
+			return
+		}
+
+		val, err := c.client.Get(ctx, key)
+		if err != nil {
+			level.Error(c.logger).Log("msg", "cleanup: failed to get replica value", "key", key, "err", err)
+			continue
+		}
+
+		desc, ok := val.(*ReplicaDesc)
+		if !ok {
+			level.Error(c.logger).Log("msg", "cleanup: got invalid replica descriptor", "key", key)
+			continue
+		}
+
+		if desc.DeletedAt > 0 {
+			if timestamp.Time(desc.DeletedAt).After(deadline) {
+				continue
+			}
+
+			// We're blindly deleting a key here. It may happen that value was updated since we have read it few lines above,
+			// in which case Distributors will have updated value in memory, but Delete will remove it from KV store anyway.
+			// That's not great, but should not be a problem. If KV store sends Watch notification for Delete, distributors will
+			// delete it from memory, and recreate on next sample with matching replica.
+			//
+			// If KV store doesn't send Watch notification for Delete, distributors *with* replica in memory will keep using it,
+			// while distributors *without* replica in memory will try to write it to KV store -- which will update *all*
+			// watching distributors.
+			err = c.client.Delete(ctx, key)
+			if err != nil {
+				level.Error(c.logger).Log("msg", "cleanup: failed to delete old replica", "key", key, "err", err)
+				c.markingOrDeletionsFailed.Inc()
+			} else {
+				level.Info(c.logger).Log("msg", "cleanup: deleted old replica", "key", key)
+				c.deletedReplicas.Inc()
+			}
+			continue
+		}
+
+		// Not marked as deleted yet.
+		if desc.DeletedAt == 0 && timestamp.Time(desc.ReceivedAt).Before(deadline) {
+			err := c.client.CAS(ctx, key, func(in interface{}) (out interface{}, retry bool, err error) {
+				d, ok := in.(*ReplicaDesc)
+				if !ok || d.DeletedAt > 0 || !timestamp.Time(desc.ReceivedAt).Before(deadline) {
+					return nil, false, nil
+				}
+
+				return &ReplicaDesc{
```

Comment:
Modifying `d.DeletedAt` instead of creating a new `ReplicaDesc` may be safer in case in the future we'll add other fields to it.

Reply:
Good point.